analysis/filter/StructureModifier/src/main/kotlin/de/maibornwolff/codecharta/filter/structuremodifier/SubProjectExtractor.kt

```diff
@@ -0,0 +1,69 @@
+package de.maibornwolff.codecharta.filter.structuremodifier
+
+import de.maibornwolff.codecharta.model.*
+import mu.KotlinLogging
+
+class SubProjectExtractor(private val project: Project, private val projectName: String?) {
+
+    private val logger = KotlinLogging.logger { }
+
+    fun extract(path: String): Project {
+        val pathSegments = path.removePrefix("/").split("/")
+        return ProjectBuilder(
+                projectName ?: project.projectName,
+                addRoot(extractNodes(pathSegments, project.rootNode.toMutableNode())),
+                extractEdges(path),
+                copyAttributeTypes(),
+                copyBlacklist()
+        ).build()
+    }
+
+    private fun extractNodes(extractionPattern: List<String>, node: MutableNode): MutableList<MutableNode> {
+        val children: List<MutableNode> = node.children
+        val extractedNodes: MutableList<MutableNode> = mutableListOf()
+
+        val currentSearchPattern = extractionPattern.firstOrNull()
+        if (currentSearchPattern == node.name && extractionPattern.size == 1) return node.children.toMutableList()
+        else if (currentSearchPattern == node.name) {
+            return children.map { child ->
+                extractNodes(extractionPattern.drop(1), child)
+            }.flatten().toMutableList()
+        }
+
+        return extractedNodes
+    }
+
+    private fun addRoot(nodes: MutableList<MutableNode>): List<MutableNode> {
+        if (nodes.size == 0) logger.warn("No nodes with the specified path(s) were fond. The resulting project is therefore empty")
+
+        val rootNode = project.rootNode.toMutableNode()
+        rootNode.children = nodes
+        return mutableListOf(rootNode)
+    }
+
+    private fun extractEdges(extractionPattern: String): MutableList<Edge> {
+            val trimmedExtractionPattern = "/" + extractionPattern.removeSuffix("/").removePrefix("/")
+        return extractRenamedEdgesForPattern(trimmedExtractionPattern).toMutableList()
+    }
+
+    private fun extractRenamedEdgesForPattern(pattern: String): List<Edge> {
+        return project.edges.filter { it.fromNodeName.startsWith(pattern) && it.toNodeName.startsWith(pattern) }
+                .map { edge ->
+                    edge.fromNodeName = "/root" + edge.fromNodeName.removePrefix(pattern)
+                    edge.toNodeName = "/root" + edge.toNodeName.removePrefix(pattern)
+                    edge
+                }
+    }
+
+    private fun copyAttributeTypes(): MutableMap<String, MutableList<Map<String, AttributeType>>> {
```

Comment:
I think this were also the correct way to collect attributeTypes, but I don't know, if that might have a high impact to the performance. `attributeList` can be `attribute types.edges` or `attributeTypes.nodes`
```
newAttributeList = attributeList.filter { attributeExistsOnAnyNode(it, newRootNode) }
```

Reply:
Theoretically you are right, we could filter the new Attribute list. However, in the visualization non-available attributes are not shown and the list of attributes will never be that long, that it significantly bloats the file size. Therefore I think filtering the attribut types (or not) has little practical impact. Or am I missing something here?

Reply:
Yea you're right ;) and the attributes are almost always the same for every node.